vmm/src/landlock.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright © 2024 Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use landlock::{
+    path_beneath_rules, Access, AccessFs, BitFlags, PathFdError, Ruleset, RulesetAttr,
+    RulesetCreated, RulesetCreatedAttr, RulesetError, ABI,
+};
+use std::io::Error as IoError;
+use std::path::PathBuf;
+use thiserror::Error;
+
+use crate::vm_config::LandlockConfig;
+
+pub const READ: u8 = 1 << 0;
+pub const WRITE: u8 = 1 << 1;
+pub const EXECUTE: u8 = 1 << 2;
+
+#[derive(Debug, Error)]
+pub enum LandlockError {
+    /// Failed to create Ruleset
+    #[error("Error creating ruleset: {0}")]
+    CreateRuleSet(#[source] RulesetError),
+
+    /// Failed to add rule
+    #[error("Error adding rule to ruleset: {0}")]
+    AddRule(#[source] RulesetError),
+
+    /// Failed to restrict self
+    #[error("Error restricting self: {0}")]
+    RestrictSelf(#[source] RulesetError),
+
+    /// Ruleset is empty
+    #[error("Ruleset was previously consumed, cannot add rules to it")]
+    RulesetConsumed,
+
+    /// Error opening Path
+    #[error("Error opening Path: {0}")]
+    OpenPath(#[source] PathFdError),
+
+    /// Error opening Device Path
+    #[error("Error opening Device Path: {0}")]
+    OpenDevicePath(#[source] IoError),
+}
+
+pub struct Landlock {
+    ruleset: Option<RulesetCreated>,
+    abi: ABI,
+}
+
+impl Landlock {
+    pub fn new() -> Result<Landlock, LandlockError> {
```

Comment:
Shouldn't `new()` without argument be replaced with `Default::default()`?

Reply:
From what I checked `default` is implemented only for `Ruleset`, which I use it in this method.

I am trying to store, a `RulesetCreated` object in `Landlock`. I don't follow how I can use `Default::default()` to replace this `new()` method.

Reply:
I was just thinking about implementing the `Default` trait for your `Landlock` struct, and then use the `default()` call name instead of `new()`. See https://rust-lang.github.io/rust-clippy/stable/index.html#/new_without_default

Reply:
@l0kod, I investigated a bit. From what I see `Default` is sensible when there are no failure cases. In this case, Landlock invokes `ruleset.create()` which could fail, if the underlying kernel does not support it. 

If I implement `Default` to Landlock, it would look like below:
```
impl Default for Landlock {
    fn default() -> Self {
        Landlock::new().unwrap()
    }
}
```
which relies on `unwrap()`, which isn't good. Please let me know if I missed anything.

Reply:
Correct, `new()` is fine.